roc/roc.go

```diff
@@ -1,170 +1,59 @@
-	return __v
-}
-
-// receiverRead function as declared in roc/receiver.h:180
-func receiverRead(receiver *Receiver, frame *frame) int32 {
-	creceiver, _ := (*C.roc_receiver)(unsafe.Pointer(receiver)), cgoAllocsUnknown
-	cframe, _ := frame.PassRef()
-	__ret := C.roc_receiver_read(creceiver, cframe)
-	__v := (int32)(__ret)
-	return __v
-}
-
-// receiverClose function as declared in roc/receiver.h:195
-func receiverClose(receiver *Receiver) int32 {
-	creceiver, _ := (*C.roc_receiver)(unsafe.Pointer(receiver)), cgoAllocsUnknown
-	__ret := C.roc_receiver_close(creceiver)
-	__v := (int32)(__ret)
-	return __v
-}
-
-// senderOpen function as declared in roc/sender.h:111
-func senderOpen(context *Context, config *SenderConfig) *Sender {
-	ccontext, _ := (*C.roc_context)(unsafe.Pointer(context)), cgoAllocsUnknown
-	cconfig, _ := config.PassRef()
-	__ret := C.roc_sender_open(ccontext, cconfig)
-	__v := *(**Sender)(unsafe.Pointer(&__ret))
-	return __v
-}
-
-// senderBind function as declared in roc/sender.h:134
-func senderBind(sender *Sender, address *Address) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	caddress, _ := address.PassRef()
-	__ret := C.roc_sender_bind(csender, caddress)
-	__v := (int32)(__ret)
-	return __v
-}
-
-// senderConnect function as declared in roc/sender.h:153
-func senderConnect(sender *Sender, kind PortType, proto Protocol, address *Address) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	ckind, _ := (C.roc_port_type)(kind), cgoAllocsUnknown
-	cproto, _ := (C.roc_protocol)(proto), cgoAllocsUnknown
-	caddress, _ := address.PassRef()
-	__ret := C.roc_sender_connect(csender, ckind, cproto, caddress)
-	__v := (int32)(__ret)
-	return __v
-}
-
-// senderWrite function as declared in roc/sender.h:177
-func senderWrite(sender *Sender, frame *frame) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	cframe, _ := frame.PassRef()
-	__ret := C.roc_sender_write(csender, cframe)
-	__v := (int32)(__ret)
-	return __v
-}
+	// ErrRuntime indicates a runtime error: memory allocation error etc
+	ErrRuntime = errors.New("Runtime error")
+)
 
-// senderClose function as declared in roc/sender.h:192
-func senderClose(sender *Sender) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	__ret := C.roc_sender_close(csender)
-	__v := (int32)(__ret)
-	return __v
+// safeString ensures that the string is NULL-terminated, a NULL-terminated copy is created otherwise.
+func safeString(str string) string {
+	if len(str) > 0 && str[len(str)-1] != '\x00' {
+		str = str + "\x00"
+	} else if len(str) == 0 {
+		str = "\x00"
+	}
+	return str
+}
+
+// NewAddress parses the `ip`, `port` and `family` and initializes the Address object
+func NewAddress(family Family, ip string, port int) (*Address, error) {
+	a := new(Address)
+	var err error
+	a.raw = (*C.roc_address)(C.calloc(1, (C.size_t)(C.sizeof_roc_address)))
```

Comment:
there is a calloc, but I suppose this is what we would expect from a ctor?

Reply:
No, we want to avoid calloc because it's not gc-collectable and requires manual free.

Instead, you can use a private byte array of size ROC_ADDRESS_SIZE. Cast it to unsafe pointer and pass to roc_address_* functions. Since they never use the pointer after returning, this is safe.

Reply:
Example of this approach: https://github.com/hraban/opus/blob/v2/encoder.go#L148

Reply:
Yeah, and one more thing. When we're passing Go memory to C and we can't be 100% sure that this memory is referenced somewhere in Go *after* the C function returns (e.g. we're holding a reference somewhere or using a variable after return), we should explicitly prolong its lifetime using runtime.KeepAlive(), to ensure that GC will not collect it until then.

Example:

```
func Foo(obj* Obj) {
    C.call_foo((C.obj*)obj)
    runtime.KeepAlive(obj) // without this call,
                  // GC can collect obj WHILE call_foo is running
}
```

OR:

```
func Foo(obj* Obj) int {
    C.call_foo((C.obj*)obj)
    return obj->x // no need for KeepAlive because we're using obj after return
}
```

If `obj` would be allocated in C memory, KeepAlive wasn't needed as well.

So we may need KeepAlive for `roc_address` and `*config` in some places. We don't need it for `roc_context`, `roc_sender`, and `roc_receiver`.